python_modules/dagster/dagster/_core/definitions/auto_materialize_rule.py

```diff
@@ -1,36 +1,101 @@
+    Sequence,
+    Set,
+    Tuple,
+    cast,
 )
 
+import dagster._check as check
 from dagster._core.definitions.data_time import CachingDataTimeResolver
 from dagster._core.definitions.events import AssetKey, AssetKeyPartitionKey
 from dagster._core.definitions.freshness_based_auto_materialize import (
-    freshness_conditions_for_asset_key,
+    freshness_evaluation_results_for_asset_key,
 )
 from dagster._core.definitions.partition_mapping import IdentityPartitionMapping
 from dagster._core.definitions.time_window_partition_mapping import TimeWindowPartitionMapping
-from dagster._serdes.serdes import whitelist_for_serdes
+from dagster._serdes.serdes import (
+    NamedTupleSerializer,
+    UnpackContext,
+    UnpackedValue,
+    WhitelistMap,
+    whitelist_for_serdes,
+)
 from dagster._utils.caching_instance_queryer import CachingInstanceQueryer
 
-from .asset_graph import AssetGraph
-from .auto_materialize_condition import (
-    AutoMaterializeCondition,
-    AutoMaterializeDecisionType,
-    MissingAutoMaterializeCondition,
-    ParentMaterializedAutoMaterializeCondition,
-    ParentOutdatedAutoMaterializeCondition,
-)
+from .asset_graph import AssetGraph, sort_key_for_asset_partition
+from .partition import SerializedPartitionsSubset
 
 if TYPE_CHECKING:
     from dagster._core.definitions.asset_daemon_context import AssetDaemonContext
     from dagster._core.definitions.asset_daemon_cursor import AssetDaemonCursor
+    from dagster._core.instance import DynamicPartitionsStore
+
+
+@whitelist_for_serdes
+class AutoMaterializeDecisionType(Enum):
+    """Represents the set of results of the auto-materialize logic.
+
+    MATERIALIZE: The asset should be materialized by a run kicked off on this tick
+    SKIP: The asset should not be materialized by a run kicked off on this tick, because future
+        ticks are expected to materialize it.
+    DISCARD: The asset should not be materialized by a run kicked off on this tick, but future
+        ticks are not expected to materialize it.
+    """
+
+    MATERIALIZE = "MATERIALIZE"
+    SKIP = "SKIP"
+    DISCARD = "DISCARD"
+
+
+class AutoMaterializeRuleEvaluationData(ABC):
+    ...
+
+
+@whitelist_for_serdes
+class GenericRuleEvaluationData(
+    AutoMaterializeRuleEvaluationData,
+    NamedTuple("_GenericRuleEvaluationData", [("description", str)]),
+):
+    ...
+
+
+@whitelist_for_serdes
+class ParentUpdatedRuleEvaluationData(
+    AutoMaterializeRuleEvaluationData,
+    NamedTuple(
+        "_ParentUpdatedRuleEvaluation",
+        [("updated_keys", FrozenSet[AssetKey]), ("will_update_keys", FrozenSet[AssetKey])],
+    ),
+):
+    ...
+
+
+@whitelist_for_serdes
+class WaitingOnParentRuleEvaluationData(
+    AutoMaterializeRuleEvaluationData,
+    NamedTuple(
+        "_WaitingOnParentRuleEvaluation",
+        [("waiting_on_keys", FrozenSet[AssetKey])],
+    ),
+):
+    ...
+
+
+@whitelist_for_serdes
+class AutoMaterializeRuleEvaluation(NamedTuple):
+    rule: "AutoMaterializeRule"
```

Comment:
Copying over my concerns here from the Notion doc: 

My general nervousness I think boils down to the fact that, up to this point, I think we don’t put any “definition-level” types onto stored events?

It seems like we’ll have to play more 4D backcompat chess in the future when we make rule changes if AutoMaterializeRules can now be present in other processes:
- The web server any time we’re deserializing a stored event.
- User writes some custom sensor to tail the the event stream of evaluations and do reporting on them.

So the question in my mind is whether the benefits we get justify this extra mental burden for us.

Reply:
Yeah definitely a reasonable concern -- my mental model for the backcompat story is that this has essentially identical constraints to any other `whitelist_for_serdes` thing that we have right now, with the added benefit of having the potential to remove the backcompat logic in the future (as the serialized evaluations with the old format can age out of the system as they're automatically deleted).

Even though definition-level events haven't been stored anywhere in the past, we still have had to deal with backcompat to allow objects serialized with an older version of dagster to be read by a newer version of dagster (this is the cloud case).

A sensor of the form you're talking about (reading from the stream of evaluations and doing something with them) would definitely cause additional constraints (as now an older version of dagster could be reading events written by a newer version), but in general, the changes we would make to these AutoMaterializeRule objects would simply be adding new fields, and the namedtuple serializer handles changes like this automatically.

So I definitely agree with you that there are additional constraints imposed by this method (as opposed to just storing strings directly), I just prefer a world in which we're able to definitively link a rule evaluation to the precise class that produced it, rather than trying to piece things together based off of raw strings. This lets us do things like make changes to the descriptions of these things and have those changes immediately propagate back to older evaluations, or have special highlighting for different types of rules, or put a tooltip in the UI saying "to remove this rule, do `policy.without_rule(AutoMaterializeRule.some_rule_name())`", stuff like that.

I'll admit none of those features are make or break, but I do feel like there is something conceptually nice with doing things this way

Reply:
> I just prefer a world in which we're able to definitively link a rule evaluation to the precise class that produced it, rather than trying to piece things together based off of raw strings

Could we address this by just having `auto_materialize_rule_repr` as one of the fields on `AutoMaterializeRuleEvaluation`?

Reply:
> This lets us do things like make changes to the descriptions of these things and have those changes immediately propagate back to older evaluations

I can definitely think of situations where this would be useful, but on balance it feels to me like more of a bug than a feature: the past should feel immutable.

Reply:
Hm ok so the concrete proposal would be:

Create a single `InternalSerializedAutoMaterializeRule` class, which contains the following attributes:

- original class name
- string description
- auto materialize decision type (skip, materialize, discard)

Then, when constructing an AutoMaterializeRuleEvaluation, we pass in this serialized rule object, rather than the actual AutoMaterializeRule. In the GQL layer, this is enough information to construct our GrapheneAutoMaterializeRule.

The main risk here is that we think of other fields we want to toss on that object in the future (e.g. if we forgot to put the decision type on it, it'd be hard to add it in, whereas the current solution it's just a property of the actual rule)